tests/notebooks/test_run_notebooks.py

```diff
@@ -1,54 +1,295 @@
 import os
 
+import nbformat
 import pytest
 from pyomo.common.fileutils import this_file_dir
 from testbook import testbook
 
 from omlt.dependencies import keras_available, onnx_available
 
+# TODO: We need to try and write these tests to rely on internal consistencies and less on absolute numbers and tolerances
 
-# TODO: These will be replaced with stronger tests using testbook soon
-def _test_run_notebook(folder, notebook_fname, n_cells):
-    # change to notebook directory to allow testing
-    cwd = os.getcwd()
+
+# return testbook for given notebook
+def open_book(folder, notebook_fname, **kwargs):
+    execute = kwargs.get("execute", True)
     os.chdir(os.path.join(this_file_dir(), "..", "..", "docs", "notebooks", folder))
-    with testbook(notebook_fname, timeout=300, execute=True) as tb:
-        assert tb.code_cells_executed == n_cells
-    os.chdir(cwd)
+    book = testbook(notebook_fname, execute=execute, timeout=300)
+    return book
+
+
+# checks that the number of executed cells matches the expected
+def check_cell_execution(tb, notebook_fname, **kwargs):
+    injections = kwargs.get("injections", 0)
+    assert (
+        tb.code_cells_executed
+        == cell_counter(notebook_fname, only_code_cells=True) + injections
+    )
+
+
+# checks for correct type and number of layers in a model
+def check_layers(tb, activations, network):
+    tb.inject(
+        f"""
+                    activations = {activations}
+                    for layer_id, layer in enumerate({network}):
+                        assert activations[layer_id] in str(layer.activation)
+                """
+    )
+
+
+# counting number of cells
+def cell_counter(notebook_fname, **kwargs):
+    only_code_cells = kwargs.get("only_code_cells", False)
+    nb = nbformat.read(notebook_fname, as_version=4)
+    nb = nbformat.validator.normalize(nb)[1]
+    if only_code_cells:
+        total = 0
+        for cell in nb.cells:
+            if cell["cell_type"] == "code" and len(cell["source"]) != 0:
+                total += 1
+        return total
+    else:
+        return len(nb.cells)
+
+
+# gets model stats for mnist notebooks
+def mnist_stats(tb, fname):
+    total_cells = cell_counter(fname)
+    tb.inject("test(model, test_loader)")

```

Comment:
What is this line doing? Can we add some more comments that describe what these (and other) lines are doing?